### PR TITLE
rewrite coprocessor substitution routine

### PR DESCRIPTION
**Pull request by gballet**
In #398 a request for a code rewrite was made by @Schaeff , to improve readability. The PR was merged before I could make the suggestion to work, so here it is.

### CODE REVIEW

File: riscv/src/compiler.rs [hunk 1]
```diff
@@ -1281,11 +1291,11 @@ fn process_instruction(instr: &str, args: &[Argument]) -> Vec<String> {
             only_if_no_write_to_zero_vec(
                 vec![
                     format!("tmp1 <== wrap({rs} + {off});"),
-                    "addr <== and(tmp1, 0xfffffffc);".to_string(),
-                    "tmp2 <== and(tmp1, 0x3);".to_string(),
-                    format!("{rd} <== mload();"),
-                    format!("{rd} <== shr({rd}, 8 * tmp2);"),
-                    format!("{rd} <== sign_extend_byte({rd});"),
+                    "addr <== and(tmp1, 0xfffffffc);".to_string(), // va lire 4 bytes alignés sur 32 bytes
```
Comment (leonardoalt) [inline riscv/src/compiler.rs:1294]: commentaires en anglais svp

Comment (gballet) [reply riscv/src/compiler.rs:1294]: oups

Comment (Schaeff) [reply riscv/src/compiler.rs]: mdr

File: riscv/src/compiler.rs [hunk 2]
```diff
@@ -307,23 +307,33 @@ fn replace_dynamic_label_reference(
...
-                to_delete.insert(i); // for the label
-                to_delete.insert(i + 1); // for the `ret` instruction
+    statements
+        .into_iter()
+        .scan(true, move |keep_next, statement| {
+            let mut keep_current = *keep_next;
+
+            // Skip the current statement and the following one if the current
+            // statement is a label that is in the list of stubs to be replaced
+            // by a corpocessor call.
```
Comment (Schaeff) [inline riscv/src/compiler.rs:325]:
```suggestion
            // by a coprocessor call.
```

File: riscv/src/compiler.rs [hunk 3]
```diff
@@ -307,23 +307,33 @@ fn replace_dynamic_label_reference(
...
+            if let Statement::Label(label) = &statement {
+                *keep_next = !stub_names.contains(&label.as_str());
+
+                keep_current &= *keep_next;
+            } else {
+                *keep_next = true;
             }
```
Comment (Schaeff) [inline riscv/src/compiler.rs:332]:
```suggestion
            if let Statement::Label(label) = &statement {
                if stub_names.contains(&label.as_str()) {
                    *keep_next = false;
                    *keep_current = false;
                }
            } else {
                *keep_next = true;
            }
```

I think something like this is clearer (I haven't checked correctness)

File: riscv/src/compiler.rs [hunk 4]
```diff
@@ -307,23 +307,34 @@ fn replace_dynamic_label_reference(
...
+                    false
+                }
+                _ => {
+                    *keep_next = true;
+                    true
+                }
+            };
+
+            Some((keep_current, statement))
+        })
```
[the remark this answers is not in the document]
Comment (gballet) [reply riscv/src/compiler.rs:336]: you're right, it's wrong... and still it works :thinking: got to dig deeper here

[the remark this answers is not in the document]
Comment (gballet) [reply riscv/src/compiler.rs:336]: ok found the issue, it was broken all along but the unfiltered line was hard to spot

[the remark this answers is not in the document]
Comment (Schaeff) [reply riscv/src/compiler.rs]: Nice! Could you add a test for that? This is quite a generic function over iterators that we may need somewhere else (I think you mentioned it appears somewhere else already?) so I think it should be extracted and tested.

File: riscv/src/compiler.rs [hunk 5]
```diff
@@ -307,6 +307,23 @@ fn replace_dynamic_label_reference(
...
+    F: Fn(&I::Item) -> bool,
+{
+    iter.scan(false, move |filter_next, item| {
+        let mut filter_current = *filter_next;
+        *filter_next = predicate(&item);
+        // if the predicate says this line should be filtered, then
+        // the next one should be filtered as well.
+        if *filter_next {
+            filter_current = true;
+        }
```
Comment (Schaeff) [inline riscv/src/compiler.rs:321]: Good catch, this is even simpler!
We can even push it to `filter_current |= *filter_next`

File: riscv/src/compiler.rs [hunk 6]
```diff
@@ -1376,3 +1377,38 @@ fn process_instruction(instr: &str, args: &[Argument]) -> Vec<String> {
...
+    use super::*;
+
+    #[test]
+    fn test_filter_matching_and_next_integers() {
+        assert_eq!(
+            filter_matching_and_next([0, 1, 2, 0, 2, 2, 1].iter(), |&&i| { i == 0 })
+                .map(|i| *i).collect::<Vec<_>>(),
+            vec![2, 2, 1]
+        );
+    }
```
Comment (Schaeff) [inline riscv/src/compiler.rs:1394]: What should happen if we have two zeroes in a row?

Comment (gballet) [reply riscv/src/compiler.rs]: they are both removed, along with the following number, which is what we want: if the label is duplicated, we still want to remove the next statement.

Comment (gballet) [reply riscv/src/compiler.rs]: added that use case to the test btw

File: riscv/src/compiler.rs [hunk 7]
```diff
@@ -307,23 +307,34 @@ fn replace_dynamic_label_reference(
...
+    F: Fn(&I::Item) -> bool,
+{
+    iter.scan(false, move |filter_next, item| {
+        let mut filter_current = *filter_next;
+        *filter_next = predicate(&item);
+        // if the predicate says this line should be filtered, then
+        // the next one should be filtered as well.
+        if *filter_next {
+            filter_current = true;
+        }
```
Comment (Schaeff) [inline riscv/src/compiler.rs:321]:
```suggestion
      filter_current |= *filter_next
```

**Review by Schaeff — APPROVED** [review]

**Pull request merged**